processor/src/main/java/org/mapstruct/ap/internal/processor/MapperCreationProcessor.java

```diff
@@ -446,41 +446,48 @@ private void mergeInheritedOptions(SourceMethod method, MapperConfiguration mapp
-            if ( applicablePrototypeMethods.size() == 1 ) {
-                mappingOptions.applyInheritedOptions(
-                    first( applicablePrototypeMethods ).getMappingOptions(),
-                    false,
-                    method,
-                    messager,
-                    typeFactory );
-            }
-            else if ( applicablePrototypeMethods.size() > 1 ) {
-                messager.printMessage(
-                    method.getExecutable(),
-                    Message.INHERITCONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH,
-                    Strings.join( applicablePrototypeMethods, ", " ) );
+        else if ( inheritanceStrategy.isAutoInherit() ) {
+
+            if ( inheritanceStrategy.isApplyForward() ) {
+                if ( applicablePrototypeMethods.size() == 1 ) {
+                    mappingOptions.applyInheritedOptions(
+                        first( applicablePrototypeMethods ).getMappingOptions(),
+                        false,
+                        method,
+                        messager,
+                        typeFactory );
+                }
+                else if ( applicablePrototypeMethods.size() > 1 ) {
+                    messager.printMessage(
+                        method.getExecutable(),
+                        Message.INHERITCONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH,
+                        Strings.join( applicablePrototypeMethods, ", " ) );
+                }
             }
 
-            if ( applicableReversePrototypeMethods.size() == 1 ) {
-                mappingOptions.applyInheritedOptions(
-                    first( applicableReversePrototypeMethods ).getMappingOptions(),
-                    true,
-                    method,
-                    messager,
-                    typeFactory );
-            }
-            else if ( applicableReversePrototypeMethods.size() > 1 ) {
-                messager.printMessage(
-                    method.getExecutable(),
-                    Message.INHERITINVERSECONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH,
-                    Strings.join( applicablePrototypeMethods, ", " ) );
+            if ( inheritanceStrategy.isApplyReverse() ) {
+                if ( applicableReversePrototypeMethods.size() == 1 ) {
+                    mappingOptions.applyInheritedOptions(
+                        first( applicableReversePrototypeMethods ).getMappingOptions(),
+                        true,
+                        method,
+                        messager,
+                        typeFactory );
+                }
+                else if ( applicableReversePrototypeMethods.size() > 1 ) {
+                    messager.printMessage(
+                        method.getExecutable(),
+                        Message.INHERITINVERSECONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH,
+                        Strings.join( applicablePrototypeMethods, ", " ) );
```

Comment:
I don't understand what you mean. The calling side would look like:

```java
else if ( inheritanceStrategy.isAutoInherit() ) {

            if ( inheritanceStrategy.isApplyForward() ) {
                applyInheritedOptions( method, mappingOptions, applicablePrototypeMethods, false );
            }

            if ( inheritanceStrategy.isApplyReverse() ) {
                applyInheritedOptions( method, mappingOptions, applicableReversePrototypeMethods, true );
            }
        }
```
Which is exactly the same. With the only difference those code that is the same is in a method now 😄